### PR TITLE
Css grid for thumbnails

### LINKED ISSUE #905: Make thumbnails grid- and flexbox-compatible
At the moment, sphinx-gallery generates a div for each gallery item.
These divs are `float: left` and have a pre-determined width, which does not render very nicely on most screen sizes or for long titles.

I suggest that for each subsection of the gallery, we add a parent div to these item divs.
This would allow users to either use css `grid` or `flex` (which we could also implement here directly).
It boils down to adding a new div before iterating through each gallery item and closing this div after:
https://github.com/sphinx-gallery/sphinx-gallery/blob/a68a48686138741ac38dcc15930293caeebcf484/sphinx_gallery/gen_rst.py#L371-L382

Also, maybe we could use [`sphinx-design`](https://github.com/executablebooks/sphinx-design) to handle the grid for us, but this would add a new dependency.

Happy to here what you think about this!

### PR DESCRIPTION
**Pull request by alexisthual**
Fixes #905.

### CODE REVIEW

**Comment by alexisthual** [comment]
I think this PR is good to go, let me know what you think :slightly_smiling_face: 

Here are some explanations to ease the code review:
- there is a new div `.sphx-glr-thumbnails` which is the parent div of all gallery items. It is a css grid (flexbox doesn't render well with `justify-content: space-between`)
- `.sphx-glr-thumbcontainer` is now light/dark theme compatible; it uses flexbox to align children horizontally
- instead of using HTML figure + figure caption to generate thumbnails, we now have img + div with text
- an HTML `a` link covers the entire `.sphx-glr-thumbcontainer` for users to click
- prettified `theme_override.css` with prettierjs :lipstick: 
- overall, this yields a shorter template for `.sphx-glr-thumbcontainer` (fewer lines), so I needed to update tests accordingly (in particular `test_full.py:test_minigallery_directive`)
- documented changes in the changelog

I tested this on a new version of nilearn's documentation and it feels nice:
Screenshot from 2022-01-18 15-11-27

**Comment by larsoner** [comment]
@lucyleeow do you want to look first?

**Review by lucyleeow — COMMENTED** [review]
I've had a look @larsoner but I am not very familiar with html/css. The parts I understood look good and I agree it looks better.
My only note is: is this comment relevant? https://github.com/sphinx-gallery/sphinx-gallery/issues/905#issuecomment-1011353892

**Comment by alexisthual** [comment]
> My only note is: is this comment relevant? [#905 (comment)](https://github.com/sphinx-gallery/sphinx-gallery/issues/905#issuecomment-1011353892)

Here are some thoughts about the comment.
This PR changes the DOM structure of thumbnails. The `<figure>` and `<figcaption>` elements are turned into an `` and a `<div>` respectively. It will most likely disable user's custom CSS although it won't throw an error :slightly_smiling_face: 

In order for users to have clear instructions on how to update their custom CSS, I think I could write a comment with all needed details in this PR, and add a link to it in the changelogs.

**Comment by larsoner** [comment]
I tested this out locally on MNE-Python's sphinx gallery page and it worked well! @alexisthual have you also tested it locally on some projects you currently maintain/contribute to? If so, +1 for merge from me, except just one more thing...

>  I think I could write a comment with all needed details in this PR, and add a link to it in the changelogs.

Indeed I think this would be valuable. Feel free to push this, and @lucyleeow and I can give it a read and merge!

**Comment by alexisthual** [comment]
> I tested this out locally on MNE-Python's sphinx gallery page and it worked well!

Yaaaay, glad to hear that :smile: 

> @alexisthual have you also tested it locally on some projects you currently maintain/contribute to?

I am actually contributing this as part of a big theme refactoring for [nilearn](https://github.com/nilearn/nilearn) (see [relevant PR](https://github.com/nilearn/nilearn/issues/3122) if you're interested). We are switching to [furo](https://github.com/pradyunsg/furo/), and this PR works great with this theme.

> Indeed I think this would be valuable. Feel free to push this,

Great, I'll do this soon!

**Comment by larsoner** [comment]
> We are switching to furo, and this PR works great with this theme.

Okay great -- MNE uses pydata-sphinx-theme so it's good to know it looks good in multiple themes

**Comment by alexisthual** [comment]
## Thumbnail CSS migration guide

In case your project includes some custom css dedicated to `sphinx-gallery` thumbnails, here are some pointer on how to update it.

### Thumbnails grid

Thumbnails are now displayed using [css grids](https://css-tricks.com/snippets/css/complete-guide-grid/) instead of [css float](https://www.w3schools.com/css/css_float.asp).
A new parent div was added to align thumbnails:
```html
<div class="sphx-glr-thumbnails">
  <div class="sphx-glr-thumbcontainer">...</div>
  <div class="sphx-glr-thumbcontainer">...</div>
  ...
</div>
```

You can tweak the grid behaviour by modifying `.sphx-glr-thumbnails` in your custom css file.

### Thumbnail

Concerning `.sphx-glr-thumbcontainer` divs, inner class names as well as overall DOM structure was modified.

Previous thumbnail DOM structure:
```html
<div class="sphx-glr-thumbcontainer" tooltip="{tooltip}">
  <div class="figure align-default">
    <img />
    <p class="caption">
      <span class="caption-text">
        <a class="reference internal">
          <span class="std std-ref">{link}</span>
        </a>
      </span>
      <a class="headerlink">¶</a>
    </p>
  </div>
</div>
```

New thumbnail DOM structure:
```html
<div class="sphx-glr-thumbcontainer" tooltip="{tooltip}">
  <img />
  <p>
    <a class="reference internal">
      <span class="std std-ref">{link}</span>
    </a>
  </p>
  <div class="sphx-glr-thumbnail-title">{link}</div>
</div>
```

Your custom css should be modified to match these new class names.

**Comment by alexisthual** [comment]
> Indeed I think this would be valuable. Feel free to push this, and @lucyleeow and I can give it a read and merge!

I just added a comment above (which can easily be edited) and a link to the changelog.

Also, I added one last change to make the grid columns responsive. Each column will be at least 160px (so that two columns can fit on smartphones), but these columns will be able to grow if more space is available (they can't grow too much though, if not it means we could add a new column).
Here are some tests on a variety of small screen sizes, I think it works well!

 
Screenshot from 2022-01-23 23-01-10
Screenshot from 2022-01-23 23-01-22
Screenshot from 2022-01-23 23-01-39

Screenshot from 2022-01-23 23-02-03

**Review by lucyleeow — COMMENTED** [review]
Thanks @alexisthual, nice write up here: https://github.com/sphinx-gallery/sphinx-gallery/pull/906#issuecomment-1019542067
I would be tempted to add this to [`faq.rst`](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/doc/faq.rst) as a section with something like 'Why has my thumbnail appearance changed?'

**Comment by alexisthual** [comment]
> I would be tempted to add this to [`faq.rst`](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/doc/faq.rst) as a section with something like 'Why has my thumbnail appearance changed?'

Done :slightly_smiling_face: The new subsection in the FAQ mainly contains a link to the migration guide and to the PR.

File: CHANGES.rst [hunk 1]
```diff
@@ -8,6 +8,22 @@ In this version, the "Out:" prefix applied to code outputs is now created from
 CSS pseudo-elements instead of additional real text. For more details, see
 `#896 <https://github.com/sphinx-gallery/sphinx-gallery/pull/896>`.
 
+v0.10.2
```
Comment (larsoner) [inline CHANGES.rst:11]: I think this will merit a 0.11.0 release, since this is a big enough change
```suggestion
v0.11.0
```

**Pull request merged**